source/language/openpulse.rst

```diff
@@ -0,0 +1,316 @@
+
+In addition to OpenQASM instructions, ``defcal`` blocks may contain OpenPulse 
+instructions. Or equivalently stated: OpenPulse instructions cannot appear
+outside of a ``defcal`` block.
+
+These instructions are motivated by the original OpenPulse specification which
+was defined as a JSON wire-format for pulse-level quantum programs in the paper
+`Qiskit Backend Specifications for OpenQASM and OpenPulse Experiment <https://arxiv.org/abs/1809.03452>`_. 
+The text format described here has several advantages over the equivalent JSON 
+format:
+
+- It is more readable
+- Absolute time is handled through built-in timekeeping instructions
+- Gates and classical instructions can be mixed in with the pulses to create far richer calibrations
+- Pulse definitions are tied to circuit instructions rather than circuit programs
+- Richer handling of frames
+
+Channels
+--------
+
+Channels map to a hardware resource, which can play pulses to manipulate a qubit
+or capture signal from the qubit to perform a measurement. There is a many-to-many
+relationship between qubits and channels. One qubit may have multiple channels
+connecting to it. Pulses on different channels would have different physical
+interactions with that qubit. A channel may also have many qubits. For instance,
+a channel could manipulate the coupling between two neighboring qubits, or
+could even reference multiple qubits in a chain.
+
+Channels are vendor-specific and device-specific. It is expected that vendors
+of quantum hardware provide the appropriate channel names and qubit mappings
+for use by end users. Each channel may also have associated static settings,
+such as local-oscillator frequencies, which do not vary throughout program
+execution. Again it is expected that vendors of quantum hardware provide a
+method for manipulating those static settings if appropriate.
+
+There are two kinds of channels: transmit channels (sending input to a quantum
+device) and receive channels (reading output from a quantum device).
+
+The ``channel`` type is a compile-time type representing a channel.
+
+A value of type ``channel`` is retreived with the special functions
+``transmit`` and ``receive``. These functions take an optional string argument
+followed by 1 or more physical qubits.
+
+The hardware vendor, during compilation, will substitute valid calls to
+``transmit`` and ``receive`` with their corresponding channels.
+
+A recommended convention: calling ``transmit`` with a single qubit argument
+should return the channel for driving single qubit gates, and calling
+``receive`` with a single qubit argument should return the channel for reading
+out the state of that qubit.
+
+.. code-block:: none
+
+   // Get the default input channel for qubit 0
+   transmit(%0)
+   // Get the default output channel for qubit 1
+   receive(%1)
+   // Get an input channel named "measure" for qubit 1
+   // Could be used to specify the measurement stimulus channel
+   transmit("measure", %1)
```

Comment:
It might be better to swap the order here to more easily define the string as an optional extension.